### PR TITLE
feat(ddm-alerts): Support metrics based alert creation

### LINKED ISSUE #59823: Alert wizard wired to metrics APIs (just /meta and /tags, /data will be in phase II)

### PR DESCRIPTION
**Pull request by ArthurKnaus**
**Note**: this is the first iteration on this feature which aim to allow to test the e2e use-case before refining it. All changes are behind the `ddm-experimental` feature flag.

Add new controls that allow selecting custom metrics for alert creation.
Hide event type selection when custom metric alerts are selected.
Load MRI specific tags and tag values in search bar.

Screenshot 2023-11-16 at 14 24 38

- closes #59823

### CODE REVIEW

**Review by matejminar — APPROVED** [review]
1. Can we remove Total Transactions + number here please?
image

2. This fails for me - maybe a backend problem, will talk to Riccardo.
image

3. I wonder how we'll solve the issue when we change the project, the new project does not have that metric and the request fails. Or when you first come to this page, what's the default metric?

Looks great!

File: static/app/views/alerts/rules/metric/mriField.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  };
+}
+
+function filterAndSortOperations(operations: string[]) {
+  return operations.filter(isAllowedOp).sort((a, b) => a.localeCompare(b));
+}
+
+function MriField({aggregate, project, onChange}: Props) {
+  const meta = useMetricsMeta([parseInt(project.id, 10)], {
+    useCases: ['transactions', 'custom'],
```
[the remark this answers is not in the document]
Comment (ArthurKnaus) [reply static/app/views/alerts/rules/metric/mriField.tsx:43]: We decided to skip this for now as it requires additional work.

**Comment by ArthurKnaus** [comment]
@matejminar 
> Can we remove Total Transactions + number here please?

Done ✅ 

> This fails for me - maybe a backend problem, will talk to Riccardo.

BE for this is still WIP

> I wonder how we'll solve the issue when we change the project, the new project does not have that metric and the request fails. Or when you first come to this page, what's the default metric?

It already falls back to the first available metric. The only problem left is if there is no available metric in the project.

File: static/app/views/alerts/rules/metric/mriField.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    [selectedMriMeta]
+  );
+
+  // As SelectControl does not support an options size limit out of the box
+  // we work around it by using the async variant of the control
+  const getMriOptions = useCallback(
+    (searchText: string) => {
+      const filteredMeta = metaArr.filter(
+        ({name}) =>
+          searchText === '' || name.toLowerCase().indexOf(searchText.toLowerCase()) > -1
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/mriField.tsx:94]:
```suggestion
          searchText === '' || name.toLowerCase().includes(searchText.toLowerCase())
```

File: static/app/views/alerts/rules/metric/ruleForm.spec.tsx [hunk 3]
```diff
@@ -234,9 +234,7 @@ describe('Incident Rules Form', () => {
       });
 
       await waitFor(() =>
-        expect(screen.getByTestId('alert-total-events')).toHaveTextContent(
-          'Total Transactions5'
-        )
+        expect(screen.getByTestId('alert-total-events')).toHaveTextContent('Total5')
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleForm.spec.tsx:237]: this should be a `await findByTestId` not a waitFor

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 4]
```diff
@@ -448,7 +456,21 @@ class RuleConditionsForm extends PureComponent<Props, State> {
                 flexibleControlStateSize
               >
                 {({onChange, onBlur, onKeyDown, initialData, value}) => {
-                  return (
+                  return hasDdmAlertsSupport(organization) ? (
+                    <MetricSearchBar
+                      mri={parseAggregate(aggregate).mri}
+                      projectIds={[parseInt(project.id, 10)]}
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:462]: can we keep proejct ids as strings? been trying to keep that consistent

File: static/app/views/ddm/queryBuilder.tsx [hunk 5]
```diff
@@ -233,6 +242,8 @@ function MetricSearchBar({tags, mri, disabled, onChange, query}: MetricSearchBar
       placeholder={t('Filter by tags')}
       query={query}
       savedSearchType={SavedSearchType.METRIC}
+      projectIds={projectIds}
```
Comment (scttcper) [inline static/app/views/ddm/queryBuilder.tsx:245]: i see the searchbar wants them as number, maybe lets convert them to number as close as possible to the searchbar

Comment (scttcper) [reply static/app/views/ddm/queryBuilder.tsx:245]: ugh
https://github.com/getsentry/sentry/blob/master/static/app/components/events/searchBar.tsx#L160

File: static/app/views/alerts/rules/metric/ruleForm.spec.tsx [hunk 6]
```diff
@@ -234,7 +234,7 @@ describe('Incident Rules Form', () => {
       });
 
       await waitFor(() =>
-        expect(screen.getByTestId('alert-total-events')).toHaveTextContent('Total5')
+        expect(screen.findByTestId('alert-total-events')).toHaveTextContent('Total5')
```
Comment (scttcper) [inline static/app/views/alerts/rules/metric/ruleForm.spec.tsx:237]: need an await here and remove the waitFor since its no longer necessary

Comment (ArthurKnaus) [reply static/app/views/alerts/rules/metric/ruleForm.spec.tsx]: ofc, thx! it is clearly too late in my time zone 🙈

**Review by scttcper — APPROVED** [review]

**Pull request merged**